### DISCUSSION TITLE
oracle 包含创建表和创建索引提示表不存在

### DISCUSSION THREAD

**Discussion by MQMMMQM** [hhyo/Archery#1697] [General]
create table testtab_a (id int primary key,aname varchar2(100));

create index idx_aname on testtab_a(aname);

insert into testtab_a values(1,'AAA');

这种需求如何解决，
我看oracle.py中有相关的 WARNING:新建表的新建索引语句暂无法检测！
但是还是提示表不存在。
![微信截图_20220721163100](https://user-images.githubusercontent.com/109661148/180168710-fc9a4cba-cbdb-4cfd-a774-e67ddb1b412b.png)

**Comment by MQMMMQM** [comment]
已自行解决。

将oracle.py 中
def execute_check 中
通过explain对SQL做语法语义检查 中
的代码

                        object_name = self.get_sql_first_object_name(sql=sql_lower)
                        if "." in object_name:
                            object_name = object_name
                        else:
                            object_name = f"""{db_name}.{object_name}"""
                        object_name_list.add(object_name)
移至try中

    def execute_check(self, db_name=None, sql="", close_conn=True):
        """
        上线单执行前的检查, 返回Review set
        update by Jan.song 20200302
        使用explain对数据修改预计进行检测
        """
        config = SysConfig()
        check_result = ReviewSet(full_sql=sql)
        # explain支持的语法
        explain_re = r"^merge|^update|^delete|^insert|^create\s+table|^create\s+index|^create\s+unique\s+index"
        # 禁用/高危语句检查
        line = 1
        # 保存SQL中的新建对象
        object_name_list = set()
        critical_ddl_regex = config.get("critical_ddl_regex", "")
        p = re.compile(critical_ddl_regex)
        check_result.syntax_type = 2  # TODO 工单类型 0、其他 1、DDL，2、DML
        try:
            sqlitemList = get_full_sqlitem_list(sql, db_name)
            for sqlitem in sqlitemList:
                sql_lower = sqlitem.statement.lower().rstrip(";")
                sql_nolower = sqlitem.statement.rstrip(";")
                object_name = self.get_sql_first_object_name(sql=sql_lower)
                if "." in object_name:
                    object_name = object_name
                else:
                    object_name = f"""{db_name}.{object_name}"""
                object_name_list.add(object_name)

**Comment by ghost** [comment]
修改这两个方法，不要加try catch这种方式
`@staticmethod
    def check_create_index_table(sql='', object_name_list=None, db_name=''):
        schema_name = ('"' + db_name + '"')
        object_name_list = object_name_list or set()
        if re.match(r"^create\s+index\s", sql):
            table_name = re.match(r"^create\s+index\s+.+\s+on\s(.+?)(\(|\s\()", sql, re.M).group(1)
            if '.' not in table_name:
                table_name = f"{schema_name}.{table_name}"
            table_name = table_name.upper()
            if table_name in object_name_list:
                return True
            else:
                return False
        elif re.match(r"^create\s+unique\s+index\s", sql):
            table_name = re.match(r"^create\s+unique\s+index\s+.+\s+on\s(.+?)(\(|\s\()", sql, re.M).group(1)
            if '.' not in table_name:
                table_name = f"{schema_name}.{table_name}"
            table_name = table_name.upper()
            if table_name in object_name_list:
                return True
            else:
                return False
        else:
            return False

    @staticmethod
    def get_dml_table(sql='', object_name_list=None, db_name=''):
        schema_name = ('"' + db_name + '"')
        object_name_list = object_name_list or set()
        if re.match(r"^update", sql):
            table_name = re.match(r"^update\s(.+?)\s", sql, re.M).group(1)
            if '.' not in table_name:
                table_name = f"{schema_name}.{table_name}"
            table_name = table_name.upper()
            if table_name in object_name_list:
                return True
            else:
                return False
        elif re.match(r"^delete", sql):
            table_name = re.match(r"^delete\s+from\s+([\w-]+)\s*", sql, re.M).group(1)
            if '.' not in table_name:
                table_name = f"{schema_name}.{table_name}"
            table_name = table_name.upper()
            if table_name in object_name_list:
                return True
            else:
                return False
        elif re.match(r"^insert\s", sql):
            table_name = re.match(r"^insert\s+((into)|(all\s+into)|(all\s+when\s(.+?)into))\s+(.+?)(\(|\s)", sql,
                                  re.M).group(6)
            if '.' not in table_name:
                table_name = f"{schema_name}.{table_name}"
            table_name = table_name.upper()
            if table_name in object_name_list:
                return True
            else:
                return False
        else:
            return False`

**Comment by LeoQuote** [comment]
麻烦提个pr吧，感谢贡献
